### ISSUE TITLE
The new general container type

### ISSUE THREAD

**Issue by drwells** [dealii/dealii#4883]
During the workshop we added a new template argument type `GENERAL_CONTAINER_TYPE`, which may be either a `std::vector` or an `ArrayView`. Part of the discussion in #4723 involves adding an implicit converting constructor from a `std::vector<T>` to an `ArrayView<T>`, so (if that goes through) we will not need to instantiate those templates and things will 'just work'. Alternatively, if we don't add the constructor, I believe that `ArrayView` is the correct argument to pass (and it is not too onerous to ask people to write `make_array_view(vec)` or `make_array_view(vec.begin(), vec.end())`).

@jppelteret @luca-heltai I believe that you two added this; what do you think of my proposal?

I really like the new functions we added; my only reservations about them are due to the fact that they now take 50% more time for me to compile.

**Comment by jppelteret** [comment]
Thanks, I've got no issues with your proposal. I think that we instantiate for both purely for compatibility with in the library, along with enhanced ease of use for users. I'm in favour of having an implicit converting constructor because it makes using these functions within user code more natural (one would certainly be using a `std::vector` there). I would then like us to explicitly note this in the documentation for `get_function_*_from_dof_values()`.

**Comment by bangerth** [comment]
Yes, I think that adding a default conversion constructor is reasonable.

**Comment by drwells** [comment]
Now that we have #5062 I gave this another look. Consider the following example:

```cpp
#include <deal.II/base/array_view.h>
#include <vector>

template <typename T>
std::size_t
foo(dealii::ArrayView<T>  u,
    std::vector<float>    &v)
{
  return u.size() + v.size();
}

int main()
{
  std::vector<double> u;
  std::vector<float> v;
  foo(u, v); // compilation error
  foo<double>(u, v); // OK
}
```

the first call to `foo` fails with the following error message:
```
[drwells@archway dealii-arrayview]$ make run
Scanning dependencies of target test
[ 33%] Building CXX object CMakeFiles/test.dir/test.cc.o
[...]/dealii-arrayview/test.cc: In function 'int main()':
[...]/dealii-arrayview/test.cc:16:11: error: no matching function for call to 'foo(std::vector<double>&, std::vector<float>&)'
   foo(u, v);
           ^
[...]/dealii-arrayview/test.cc:6:1: note: candidate: template<class T> std::size_t foo(dealii::ArrayView<ElementType>, std::vector<float>&)
 foo(dealii::ArrayView<T>  u,
 ^~~
[...]/dealii-arrayview/test.cc:6:1: note:   template argument deduction/substitution failed:
[...]/dealii-arrayview/test.cc:16:11: note:   'std::vector<double>' is not derived from 'dealii::ArrayView<ElementType>'
   foo(u, v);
           ^
Command exited with non-zero status 1
```

I believe that the compiler does not include conversion constructors when trying to figure out `T`. This means that our intended use case for #5062 is not going to happen: we cannot write


```cpp
    template <typename Number>
    void get_function_values_from_local_dof_values (const ArrayView<const Number> dof_values,
                                                    std::vector<typename OutputType<Number>::value_type> &values) const;
```

and pass the method a `std::vector` as the first argument and let the compiler figure out the conversion constructor.

Hence, the converting constructor doesn't really help us that much since we cannot use it as an implicit conversion for arguments to template functions (e.g., virtually everything we have). I don't see a good way around this besides making everyone write `make_array_view` hundreds of times (not that I am opposed to that solution, but we all already know that :)). Thoughts?

(this test was done with GCC 7; I get a similar result with clang 4)

**Comment by bangerth** [comment]
You can make it happen if you make the argument a `const` reference to an `ArrayView`.

**Comment by drwells** [comment]
Yup, that does it.

I am surprised that it works for const references and not values. Do you see why this should be the case?

**Comment by bangerth** [comment]
I don't have chapter and verse ready. As you did, I would have expected that a const reference or a value argument both yield the same conversion. Apparently we're wrong. It's clear, though, that a non-`const` reference would not have worked because that can't bind to the rvalue you get from the conversion.

By the way, I have a first version of a single function converted. Give me an hour while my machine churns through the testsuite to post that version.

**Comment by bangerth** [comment]
See #5100.
